strutil/quantity/quantity.go

```diff
@@ -200,3 +201,247 @@ func FormatDuration(dt float64) string {
 }
+
+// Duration is represented in nano seconds.
+type Duration uint64
+
+const (
+	NSec Duration = 1
+	USec Duration = 1000
+	MSec Duration = 1000000
+	Sec  Duration = 1000000000
+	Min  Duration = (60 * Sec)
+	Hour Duration = (60 * Min)
+	Day  Duration = (24 * Hour)
+	Year Duration = (1461 * Day / 4)
+)
+
+// Maximum number of places (units of time) to render starting with years.
+// In the Compact case, only the first two places will be rendered,
+// discarding the rest. This allows for a more predictable width (at the
+// expense of accuracy).
+type Places uint32
+
+const (
+	ShowAll  Places = 8
+	ShowComp Places = 2
+)
+
+// Space delimit options
+type SpaceMode bool
+
+const (
+	SpaceOff SpaceMode = false
+	SpaceOn  SpaceMode = true
+)
+
+// The minimum Duration parameter allows the rendering to discard the units
+// of time less than the minimum. However, depending on what we are
+// rendering, we either need to peform a ceiling, floor or rounding
+// operation with the remainder below the minimum.
+type RenderMode uint32
+
+const (
+	TimeLeft    RenderMode = 0
+	TimePassed  RenderMode = 1
+	TimeRounded RenderMode = 2
+)
+
+// FormatDurationGeneric formats time duration similar to systemd
+// format_timespan, but with options allowing for a more compact
+// arrangement, including shortened suffixes. The function is provided
+// with dt in seconds (as in the output of time.Now().Seconds()). The 'min'
+// and 'max' Duration allows limiting the unit range rendered. Place values
+// less than 'min' will be processed according to RenderMode. RenderMode
+// allows rounding, flooring or ceiling to be applied to place value below
+// 'min'. Durations above the 'max' place value will result in "ages!".
+func FormatDurationGeneric(dt float64, min Duration, max Duration, count Places, mode RenderMode, space SpaceMode) string {
+	var units = map[Duration]string{
+		Year: "y",
+		Day:  "d",
+		Hour: "h",
+		Min:  "m",
+		Sec:  "s",
+		MSec: "ms",
+		USec: "µs",
+		NSec: "ns",
+	}
+
+	// We need to access the map in order using an ordered array
+	var keys []Duration
+	for k := range units {
+		keys = append(keys, k)
+	}
+	sort.Slice(keys, func(i, j int) bool { return keys[i] > keys[j] })
+
+	var render string
+
+	// Invalid case: min > max
+	if min > max {
+		return "inv!"
+	}
+
+	// Special case: zero duration
+	if dt <= 0 {
+		return "0" + units[min]
+	}
+
+	// Special case: render as "ages!"
+	if dt >= float64(math.MaxUint64) {
+		return "ages!"
+	}
+
+	// Fractional seconds to nanoseconds
+	delta := uint64(dt * float64(Sec))
+
+	// If we only render a subset of place values, we need to apply
+	// the render mode to the last place value. We do this indirectly
+	// by moving up the 'min' accuracy.
+	if count == ShowComp {
+		for i := 0; i < len(keys); i++ {
+			// Only render places as indicated by count
+			if delta >= uint64(keys[i]) && keys[i] >= Min {
+				index := i + int(count) - 1
+				if keys[index] > min {
```

Comment:
I made a couple of changes to make this clearer:

(1) Changed 'Min' duration to 'Minute' so its not confused with the parameter 'min'
(2) Changed 'i + int(count) - 1' to 'i+1' as the entire block is only for ShowCompact mode (count == 2)

The code is safe because the block will only enter for durations > seconds, which means plenty of places (s, ms, us, ns) are left for +1 to always be safe on the index.